lib/http/client.rb

```diff
@@ -70,9 +70,16 @@ def perform(req, options)
 
       @connection ||= HTTP::Connection.new(req, options)
 
-      unless @connection.failed_proxy_connect?
-        @connection.send_request(req)
-        @connection.read_headers!
+      begin
+        unless @connection.failed_proxy_connect?
+          @connection.send_request(req)
+          @connection.read_headers!
+        end
+      rescue Error => e
+        options.features.each do |_name, feature|
```

Comment:
I think you can do `each_value` here (and potentially collapse it into a 1-liner), e.g.:

`options.features.each_value { |feature| feature.on_error(req, e) }`